docs/proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,106 @@
+
+##### Cloud Provider in Kubelet
+Moving on to the kubelet, the following cloud provider dependencies exist in kubelet.
+
+ - Find the cloud nodename of the host that kubelet is running on for two reasons :
+      1. To obtain the config map for the kubelet, if one already exists
+      2. To uniquely identify current node using nodeInformer
+      3. To instantiate a reference to the current node object
+ - Find the InstanceID, ProviderID, ExternalID, Zone Info of the node object while initailizing it
+ - Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it 
+ - In case of Google cloud, it allows the provider to configure routes for the node
+ - It allows the cloud provider to post process DNS settings
+ 
+##### Cloud Provider in Kube-apiserver
+ 
+Finally, in the kube-apiserver, the cloud provider is used for trasferring SSH keys to all of the nodes, and within an admission controller for setting labels on persistent volumes.
+ 
+### 2. Startegy for refactoring Kube-Controller-Manager 
+
+In order to create a 100% cloud independent controller manager, the idea is to split the controller-manager into two separate binaries. 
+
+1. Cloud dependent controller-manager binary
+2. Cloud independent controller-manager binary
+
+The cloud dependent binary will run those loops that rely on cloudprovider. The rest of the controllers will be run in the cloud independent controller manager. The decision to run entire controller loops, rather than only the very minute parts that rely on cloud provider makes implementation simple. If drilled down further into the controller loops, it would be significantly harder to disentangle, due to tight coupling of the cloud provider specific code with the rest of the data structures/functions within a controller.
+
+There are four controllers that rely on cloud provider specific code. These are node controller, service controller, route controller and attach detach controller. Copies of each of these controllers have been bundled them together into one binary. This binary registers itself as a controller, and runs the cloud specific controller loops with the user-agent named "external-controller-manager".
+
+RouteController and serviceController are entirely cloud specific. Therefore, it is really simple to run these two controller loops separately. 
+
+NodeController does a lot more than just talk to the cloud. Therefore, this can be split up into two different controllers. 
+
+Finally, The attachDetachController is tricky, and it is not simple to disentangle it from the controller-manager easily, therefore, this will be addressed with Flex Volumes (Discussed under a separate section below)
+
+### 3. Strategy for refactoring Kubelet 
+
+The majority of the calls by the kubelet to the cloud is done during the initialization of the Node Object. The other uses are for configuring Routes (in case of GCE), scrubbing DNS, and periodically polling for IP addresses. 
```

Comment:
@roberthbailey https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network.go#L85 
which is called from ->
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L189